### PR TITLE
Kill advanced settings global singleton

### PR DESCRIPTION
**Pull request by ksooo**
After updating to macOS 10.14 I experienced crashes in `CAdvancedSettings`dtor while exiting Kodi and thought it would be a good time to eliminate another global singleton: `g_advancedSettings`.

Example error message in console log:

```
[...]
2018-09-26 16:34:26.521365+0200 kodi.bin[26047:1536643] Debug Print: application stopped...
kodi.bin(26047,0x107f075c0) malloc: tiny_free_list_remove_ptr: Internal invariant broken (prev ptr of next): ptr=0x10959c290, next_prev=0x0
kodi.bin(26047,0x107f075c0) malloc: *** set a breakpoint in malloc_error_break to debug
```

Example crash:

```
(lldb) thread backtrace
* thread #1, queue = 'com.apple.main-thread', stop reason = signal SIGABRT
    frame #0: 0x00007fff79b09b86 libsystem_kernel.dylib`__pthread_kill + 10
    frame #1: 0x0000000109290884 libsystem_pthread.dylib`pthread_kill + 285
    frame #2: 0x00007fff79a731c9 libsystem_c.dylib`abort + 127
    frame #3: 0x00007fff79b816e2 libsystem_malloc.dylib`malloc_vreport + 545
    frame #4: 0x00007fff79b9586c libsystem_malloc.dylib`malloc_zone_error + 184
    frame #5: 0x00007fff79b7e130 libsystem_malloc.dylib`tiny_free_list_remove_ptr + 589
    frame #6: 0x00007fff79b7baee libsystem_malloc.dylib`tiny_free_no_lock + 933
    frame #7: 0x00007fff79b7b631 libsystem_malloc.dylib`free_tiny + 483
  * frame #8: 0x0000000100121b09 kodi.bin`std::__1::__vector_base<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > >::~__vector_base() [inlined] std::__1::__libcpp_deallocate(__ptr=0x000000010959c740) at new:236
    frame #9: 0x0000000100121b00 kodi.bin`std::__1::__vector_base<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > >::~__vector_base() [inlined] std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >::deallocate(this=0x000000010a003e88, __p="[\\/].+\\.ite[\\/]", (null)=4) at memory:1796
    frame #10: 0x0000000100121af8 kodi.bin`std::__1::__vector_base<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > >::~__vector_base() [inlined] std::__1::allocator_traits<std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > >::deallocate(__a=0x000000010a003e88, __p="[\\/].+\\.ite[\\/]", __n=4) at memory:1555
    frame #11: 0x0000000100121ae0 kodi.bin`std::__1::__vector_base<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > >::~__vector_base(this=0x000000010a003e78) at vector:442
    frame #12: 0x0000000100121935 kodi.bin`std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > >::~vector(this=0x000000010a003e78 size=0) at vector:447
    frame #13: 0x00000001001213a5 kodi.bin`std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > >::~vector(this=0x000000010a003e78 size=0) at vector:447
    frame #14: 0x0000000102183574 kodi.bin`CAdvancedSettings::~CAdvancedSettings(this=0x000000010a003c00) at AdvancedSettings.h:104
    frame #15: 0x00000001021813d5 kodi.bin`CAdvancedSettings::~CAdvancedSettings(this=0x000000010a003c00) at AdvancedSettings.h:104
    frame #16: 0x00000001021813f9 kodi.bin`CAdvancedSettings::~CAdvancedSettings(this=0x000000010a003c00) at AdvancedSettings.h:104
    frame #17: 0x000000010000b935 kodi.bin`std::__1::__shared_ptr_pointer<CAdvancedSettings*, std::__1::default_delete<CAdvancedSettings>, std::__1::allocator<CAdvancedSettings> >::__on_zero_shared() [inlined] std::__1::default_delete<CAdvancedSettings>::operator(this=0x0000000109405d08, __ptr=0x000000010a003c00)(CAdvancedSettings*) const at memory:2285
    frame #18: 0x000000010000b90c kodi.bin`std::__1::__shared_ptr_pointer<CAdvancedSettings*, std::__1::default_delete<CAdvancedSettings>, std::__1::allocator<CAdvancedSettings> >::__on_zero_shared(this=0x0000000109405cf0) at memory:3586
    frame #19: 0x000000010000bb91 kodi.bin`std::__1::shared_ptr<CAdvancedSettings>::~shared_ptr() [inlined] std::__1::__shared_count::__release_shared(this=0x0000000109405cf0) at memory:3490
    frame #20: 0x000000010000bb47 kodi.bin`std::__1::shared_ptr<CAdvancedSettings>::~shared_ptr() [inlined] std::__1::__shared_weak_count::__release_shared(this=0x0000000109405cf0) at memory:3532
    frame #21: 0x000000010000bb47 kodi.bin`std::__1::shared_ptr<CAdvancedSettings>::~shared_ptr(this=0x000000010940e260) at memory:4468
    frame #22: 0x000000010000aae5 kodi.bin`std::__1::shared_ptr<CAdvancedSettings>::~shared_ptr(this=0x000000010940e260) at memory:4466
    frame #23: 0x000000010000bafc kodi.bin`xbmcutil::GlobalsSingleton<CAdvancedSettings>::Deleter<std::__1::shared_ptr<CAdvancedSettings> >::~Deleter(this=0x00000001050f1ff0) at GlobalsHandling.h:117
    frame #24: 0x000000010000af85 kodi.bin`xbmcutil::GlobalsSingleton<CAdvancedSettings>::Deleter<std::__1::shared_ptr<CAdvancedSettings> >::~Deleter(this=0x00000001050f1ff0) at GlobalsHandling.h:117
    frame #25: 0x00007fff79a73ef2 libsystem_c.dylib`__cxa_finalize_ranges + 351
    frame #26: 0x00007fff79a741de libsystem_c.dylib`exit + 55
    frame #27: 0x00007fff799cb08c libdyld.dylib`start + 8
(lldb) 
```
For me, this actually (and as expected) fixes the crashes.

I've put the changes into two commits. First commit contains the changes needed to get rid of the global singleton, and just redefines g_advancedSettings. Second commit contains all the changes needed to actually eliminate all occurrences of g_advancedSettings from the source code.

Changes are runtime-tested on macOS, latest Kodi master and Jenkins is happy as well.

Although this is a larger code change, I suggest to include it for v18.

@MilhouseVH could you include this PR into your next testbuilds, please?

### CODE REVIEW

**Comment by MilhouseVH** [comment]
@ksooo I can include this for Generic easily enough (once the `--standalone` crash is resolved of course).

**Review by FernetMenta — DISMISSED** [review]
ServiceManager was an intermediate solution that shows how bad things are and what we need to improve. ServiceManager is not a container for any kind of services. In the meantime we developed better strategies.
Have a look at GUIComponent which encapsulates other objects like StereoscopicManager or WindowManager. We don't need advanced settings and settings at the top level.
Services are supposed to register in ServiceBroker when they become available. Check out CAppPort or CAnnouncementManager.

**Comment by ksooo** [comment]
Lets do one step after the after @FernetMenta 
This is all I can do for the moment to fix the crashes
You know me. If I say I will fix it later (v19) than you can take me serious.

**Comment by ksooo** [comment]
> We don't need advanced settings and settings at the top level.

It's accessed from >150 source files and very early at boot. This is what I call a real toplevel service.

**Comment by ksooo** [comment]
@MilhouseVH I updated the PR with a fix for the `--standalone` crash.

**Comment by FernetMenta** [comment]
> Advanced settings is accessed from >150 source files and very early at boot. This is what I call a real toplevel service.

this is also true for CWindowManager which is encapsulated in GUIComponent. It makes sense to group all settings related services into a single settings component.

**Comment by FernetMenta** [comment]
I am just saying what we learned over the past couple of months and how this could apply here. No need to follow my advice.

**Comment by ksooo** [comment]
> No need to follow my advice.

Oh, I guess I will. I just need to find out whether this is high effort. If so, I suggest to go with the current approach for v18 and to make the bigger changes for v19.

**Comment by ksooo** [comment]
jenkins build this please

File: xbmc/AppParamParser.h [hunk 1]
```diff
@@ -8,22 +8,30 @@
 
 #pragma once
 
+#include <string>
```
Comment (Rechi) [inline xbmc/AppParamParser.h]: Include order is own headers before C(++) and system headers, see https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#6-headers

File: xbmc/Application.cpp [hunk 2]
```diff
@@ -3355,21 +3354,23 @@ void CApplication::OnPlayerCloseFile(const CFileItem &file, const CBookmark &boo
 
   percent = bookmark.timeInSeconds / bookmark.totalTimeInSeconds * 100;
 
-  if ((fileItem.IsAudio() && g_advancedSettings.m_audioPlayCountMinimumPercent > 0 &&
-       percent >= g_advancedSettings.m_audioPlayCountMinimumPercent) ||
-      (fileItem.IsVideo() && g_advancedSettings.m_videoPlayCountMinimumPercent > 0 &&
-       percent >= g_advancedSettings.m_videoPlayCountMinimumPercent))
+  CAdvancedSettings& advancedSettings = CServiceBroker::GetAdvancedSettings();
```
Comment (garbear) [inline xbmc/Application.cpp]: can this be `const CAdvancedSettings& advancedSettings`? I try to take advantage of these efforts to sprinkle some const-correctness around

File: xbmc/ServiceManager.cpp [hunk 3]
```diff
@@ -250,6 +260,14 @@ void CServiceManager::DeinitStageOne()
...
   m_XBPython.reset();
 #endif
+
+  m_advancedSettings->Clear();
+  m_advancedSettings.reset();
+}
+
+CAdvancedSettings &CServiceManager::GetAdvancedSettings()
+{
+  return *m_advancedSettings.get();
```
Comment (garbear) [inline xbmc/ServiceManager.cpp]: This can be just `return *m_advancedSettings;`

**Comment by garbear** [comment]
These kind of improvements take a lot of time and effort. Thank you!

File: xbmc/Application.cpp (hunk 2 above)
Comment (ksooo) [reply xbmc/Application.cpp]: sure, will change all occurences, thanks

File: xbmc/ServiceManager.cpp (hunk 3 above)
Comment (ksooo) [reply xbmc/ServiceManager.cpp]: yep

File: xbmc/AppParamParser.h (hunk 1 above)
Comment (ksooo) [reply xbmc/AppParamParser.h]: sure, thanks

**Comment by ksooo** [comment]
jenkins build this please

**Comment by ksooo** [comment]
Super seeded by #14511

**Pull request closed without merging**